source/elements/oneDPL/source/parallel_api/parallel_range_api.rst

```diff
@@ -0,0 +1,323 @@
+defined by the C++ standard in ``namespace std::ranges``, they cannot be found by argument-dependent name lookup
+and cannot be called with explicitly specified template arguments. [*Note*: A typical implementation uses
+predefined function objects which static function call operators have the required signatures. -- *end note*]
+
+The following differences to the standard C++ range algorithms apply:
+
+- The execution policy parameter is added.
+- Output data sequences are defined as ranges, not iterators.
+- Both input and output ranges must support random access.
+- For a given algorithm, at least one of the input ranges as well as the output range must be bounded.
+- ``for_each`` does not return its function object.
+
+Whole Sequence Operations
++++++++++++++++++++++++++
+
+.. code:: cpp
+
+  // Defined in <oneapi/dpl/algorithm>
+
+  namespace oneapi::dpl::ranges {
+
+    // all_of
+    template <typename ExecutionPolicy, std::ranges::random_access_range R,
+              typename Proj = std::identity,
+              std::indirect_unary_predicate< std::projected<std::ranges::iterator_t<R>, Proj> > Pred>
+      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
+               std::ranges::sized_range<R>
+      bool all_of (ExecutionPolicy&& pol, R&& r, Pred pred, Proj proj = {});
+
+    // any_of
+    template <typename ExecutionPolicy, std::ranges::random_access_range R,
+              typename Proj = std::identity,
+              std::indirect_unary_predicate< std::projected<std::ranges::iterator_t<R>, Proj> > Pred>
+      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
+               std::ranges::sized_range<R>
+      bool any_of (ExecutionPolicy&& pol, R&& r, Pred pred, Proj proj = {});
+
+    // none_of
+    template <typename ExecutionPolicy, std::ranges::random_access_range R,
+              typename Proj = std::identity,
+              std::indirect_unary_predicate< std::projected<std::ranges::iterator_t<R>, Proj> > Pred>
+      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
+               std::ranges::sized_range<R>
+      bool none_of (ExecutionPolicy&& pol, R&& r, Pred pred, Proj proj = {});
+
+    // for_each
+    template <typename ExecutionPolicy, std::ranges::random_access_range R,
+              typename Proj = std::identity,
+              std::indirectly_unary_invocable< std::projected<std::ranges::iterator_t<R>, Proj> > Fn>
+      requires oneapi::dpl::is_execution_policy_v<std::remove_cvref_t<ExecutionPolicy>> &&
+               std::ranges::sized_range<R>
+      std::ranges::borrowed_iterator_t<R>
+        for_each (ExecutionPolicy&& pol, R&& r, Fn f, Proj proj = {});
+
+    // count
+    template <typename ExecutionPolicy, std::ranges::random_access_range R,
+              typename Proj = std::identity,
+              typename T = std::projected_value_t<std::ranges::iterator_t<R>, Proj>>
```

Comment:
It looks like we are matching the C++26 version of this API with the template ordering and by including this default (though with the mention that the algorithms can't use explicitly specified  template arguments the ordering shouldn't matter).

For my own understanding, when can this default argument for `T` be used?  Won't this type always be deduced from the supplied function argument?


Also, for this PR, should we describe that in at least some cases, we are using APIs from versions beyond C++20?

Reply:
> For my own understanding, when can this default argument for T be used?

It is used when a braced init list is passed as an argument and so type deduction is not possible. As far as I remember, C++26 adds it to all standard algorithms, and also there is a GH issue to consider doing it for oneDPL algorithms.

> Also, for this PR, should we describe that in at least some cases, we are using APIs from versions beyond C++20?

Yes, perhaps the introduction needs to be slightly rephrased in that regard.

Reply:
That makes sense. Thanks.  

I'll leave this unresolved for now to represent the comment about versions of the ranges API from beyond C++20.

Reply:
Added a sentence about the signatures matching the current working draft for C++26.